### ISSUE TITLE
Implement Bitswap protocol using notifications protocol

### ISSUE THREAD

**Issue by altonen** [paritytech/substrate#12850]
Bitswap protocol cannot be served over libp2p's request response protocol because it is not compatible with how go-ipfs and js-ipfs work. They seem to expect the response to be received in a new substream.

After #5685 has been implemented, rewrite the Bitswap protocol using notifications protocol and use the Bitswap query result as the handshake of the stream as is done in [`ipfs-embed`'s implementation](https://github.com/ipfs-rust/libp2p-bitswap/blob/master/src/compat/protocol.rs#L61-L65)

Related to #7915

**Comment by altonen** [comment]
Some relevant discussion in the libp2p repository: https://github.com/libp2p/rust-libp2p/issues/2632

Before starting the work on this, we probably need to reevaluate if this is something we absolutely want to support.

**Comment by burdges** [comment]
If I understand, bitswap is basically bittorrent over libp2p TCP connections.  It avoids some legacy bitorrent cruft, but also stupidly misses major bittorrent advancements like [uTP](http://bittorrent.org/beps/bep_0029.html) adopting UDP.

At the same time uTP exists for [LEDBAT](https://en.wikipedia.org/wiki/LEDBAT) aka lower bittorrent data priority, not necessarily our goal.  We might however want lower priority for transporting blocks and parablock chunks however since we've some quite sensitive messages which should take higher priority.

We could run a race between some open source based upon [libtorrent](https://github.com/arvidn/libtorrent) that usess uTP and a stripped down client based upon libp2p's bitswap.  It'll be funny when libtorrent smokes libp2p's bitswap despite being handicapped by LEDBAT. 

I've no idea if we want this short- to mid-term but long-term one crazy idea goes..

We should eventually figure out if outgoing UDP packets fair better under DoS attacks.  _If so_ then we could maybe send approval announcements over UDP connections, which we'd open between all validators using the existing TCP key exchange, except then UDP messages sticks out.  I'd become more interesting if we exchange blocks and parablock chunks using uTP, and approval announcements then disguise themselves as requests or whatever.  

At the extreme, we've fairly serious criticism of libp2p by @tomaka and others, but if one did a clean rust bittorrent crate including uTP, which achieved libtorrent like performance (good luck lol), then one could consider designing a p2p layer for substrate and polkadot that more closely matched and exploited the primitives in bittorrent, and ripping out libp2p.  At least we should fund someone doing a rust bittorrent crates which pays close attention to the libtorrent code.

I've not answered your question here, but maybe this is informative.

**Comment by tomaka** [comment]
>  Anyways, if libp2p abstraction break bitswap then maybe those abstractions suck and one needs more holes punched through them.

The abstractions provided by Substrate (notifications and request-responses) have been designed the way they are because they can be made DoS-resilient (they are not really right now, but it's an implementation issue, and the protocol isn't flawed).

The bitswap protocol is not DoS-resilient, and thus can't be implemented using the abstractions that we have at the moment.

But I'd argue that the solution here isn't to allow implementing non-DoS-resilient protocols.

**Comment by sinkingsugar** [comment]
> > Anyways, if libp2p abstraction break bitswap then maybe those abstractions suck and one needs more holes punched through them.
> 
> The abstractions provided by Substrate (notifications and request-responses) have been designed the way they are because they can be made DoS-resilient (they are not really right now in Substrate (unless it has changed recently but I don't think so), but it's an implementation issue, and the protocol isn't flawed).
> 
> The bitswap protocol is not DoS-resilient, and thus can't be implemented using the abstractions that Substrate has at the moment. At least that's how it was the last time I checked out the situation. I'm not really familiar with what people are talking about in the libp2p issue that has been linked.
> 
> But I'd argue that the solution here isn't to allow implementing non-DoS-resilient protocols.

No problem with that and good points. 
But that's why it is behind the `--ipfs-server` flag though. Also DoS resistance should be probably deployed at another level (Substrate doesn't need to do everything), especially RPC nodes should be behind LoadBalancers if deployed in production and to the mass I'd argue.

**Comment by tomaka** [comment]
> But that's why it is behind the --ipfs-server flag though.

No problem with that, but then this flag should be renamed something like "--ipfs-server-note-that-a-malicious-peer-could-crash-your-node", as you can't ask people to be magically aware that using a certain flag is dangerous.

> especially RPC nodes should be behind LoadBalancers if deployed in production and to the mass I'd argue.

There's no load balancer capable of handling libp2p/Substrate traffic. Also, you can't ask a load balancer to be DoS resilient while handling a protocol that can't be made DoS resilient. They're not magic.

**Comment by sinkingsugar** [comment]
> > But that's why it is behind the --ipfs-server flag though.
> 
> 
> 
> No problem with that, but then this flag should be renamed something like "--ipfs-server-note-that-a-malicious-peer-could-crash-your-node", as you can't ask people to be magically aware that using a certain flag is dangerous.
> 
> 
> 
> > especially RPC nodes should be behind LoadBalancers if deployed in production and to the mass I'd argue.
> 
> 
> 
> There's no load balancer capable of handling libp2p/Substrate traffic. Also, you can't ask a load balancer to be DoS resilient while handling a protocol that can't be made DoS resilient. They're not magic.

Oh yeah but gossipsub has its ways to resist. I was just referring about http and ws endpoints not libp2p endpoints. 
You are right about bitswap, but well, then identity, multistream and whatnot are also fairly open to abuse if done right...

Btw I've seen setups with gossip port going thru nginx (stream) and what not, not sure how well.
